File: classrooms.py
from sys import stdin
import bisect


def compatible(activity, classroom):
    if classroom == 0:
        return True
    # Activity must start after last class
    return classroom < activity[1] 

def bubble(i, array):

    while i < len(array) - 1 and array[i + 1] < array[i]:
        array[i], array[i + 1] = array[i + 1], array[i]
        i += 1
    while i > 0 and array[i - 1] > array[i]:
        array[i], array[i - 1] = array[i - 1], array[i]
        i -= 1
    



fast_input =  lambda : stdin.readline().strip()

num_activities, num_classrooms = map(int, fast_input().split())

classrooms = [0] * num_classrooms


activities = []
for i in range(num_activities):
    s_i, f_i = map(int, fast_input().split())
    activities.append((f_i, s_i))
    

activities.sort()
num_possible_act = 0


for activity in activities:
    i = bisect.bisect_left(classrooms, activity[1]) - 1
    if compatible(activity, classrooms[i]):
        classrooms[i] = activity[0]
        bubble(i, classrooms)
        num_possible_act += 1

print(num_possible_act)

File: classrooms.cpp
#include <iomanip>
#include <iostream>
#include <set>
#include <algorithm> 

using namespace std;

struct Activity{
    int start, end;

    bool operator < (const Activity &other) const {
        if(this->end == other.end){ 
            return this->start < other.start;
        }
        return this->end < other.end;
    }
};

void bubble(int i, int classrooms[], int len){
    while(i < len - 1 && classrooms[i + 1] < classrooms[i]){
        int temp = classrooms[i + 1];
        classrooms[i + 1] = classrooms[i];
        classrooms[i] = temp;
    }
     while(i > 0 && classrooms[i - 1] > classrooms[i]){
        int temp = classrooms[i - 1];
        classrooms[i - 1] = classrooms[i];
        classrooms[i] = temp;
    }

}
int main(){
    int num_activities, num_classrooms;
    cin >> num_activities >> num_classrooms;

    int classrooms[num_classrooms];

    Activity activities[num_activities];
    for(int i = 0; i < num_activities; i++){
        struct Activity a;
        cin >> a.start >> a.end;
    }

    sort(activities, activities + num_activities);

    int num_possible_act = 0;

    for(int i = 0; i < num_activities; i++){
        int k = lower_bound(classrooms, classrooms + num_classrooms, activities[i].start) - classrooms;
        if (classrooms[k] < activities[k].start){
            classrooms[k] = activities[k].end;
            bubble(i, classrooms, num_classrooms);
            num_possible_act += 1;
        }
    }

    cout << num_possible_act << endl;
    
    return 0;
}